hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,951 @@
+   * adjusting internal refs since they have correct PC relative offsets and
+   * immediate already.
+   */
+  try {
+    // Simple relocation just copies src instructions to dest instructions.
+    // If a src instruction(s) are actually rewritten, then those src
+    // instructions are tracked in this list.
+    InstrSet rewrites;
+
+    // Find the literals embedded within the range
+    InstrSet literals = findLiterals(Instruction::Cast(start),
+                                     Instruction::Cast(end));
+
+    // Relocate each instruction to the destination.
+    while (src != end) {
+      assertx(src < end);
+
+      // Initially copy the instruction word
+      auto dest = dest_block.frontier();
+      dest_block.bytes(kInstructionSize, src);
+      size_t src_count = 1, dest_count = 1;
+
+      // If it's not a literal, then attempt any special relocations
+      if (!literals.count(Instruction::Cast(src)) &&
+          !relocateSmashableHelper(rel, meta, dest_block,
+                                   start, end, src, dest, jmp_dest,
+                                   src_count, dest_count, far, rewrites,
+                                   internal_refs_need_update) &&
+          !relocatePCRelativeHelper(dest_block,
+                                    start, end, src, dest, jmp_dest,
+                                    src_count, dest_count, far, rewrites,
+                                    internal_refs_need_update) &&
+          !relocateImmediateHelper(rel, dest_block, end,
+                                   src, dest,
+                                   src_count, dest_count, rewrites,
+                                   internal_refs_need_update)) {
+        // Do nothing, as the instruction word was initially copied above
+      }
+
+      if (src == start) {
+        /*
+         * For the start of the range, we only want to overwrite the "after"
+         * address (since the "before" address could belong to the previous
+         * tracelet, which could be being relocated to a completely different
+         * address. recordRange will do that for us, so just make sure we
+         * have the right address setup.
+         */
+        dest_start = dest;
+      } else {
+        rel.recordAddress(src, dest, 0);
+      }
+
+      src += src_count * kInstructionSize;
+      dest += dest_count * kInstructionSize;
+      asm_count += dest_count;
+      assertx(dest <= dest_block.frontier());
+      dest_block.setFrontier(dest);
+    } // while (src != end)
+
+    if (exit_addr) {
+      *exit_addr = jmp_dest;
+    }
+
+    rel.recordRange(start, end, dest_start, dest_block.frontier());
+
+    /*
+     * Finally update any internal refs if needed. This indicates that the
+     * range of instructions grew/shrank and therefore the internal refs
+     * may be off.
+     */
+    if (internal_refs_need_update) {
+      src = start;
+      bool ok = true;
+      while (src != end) {
```

Comment:
Can you make into for loops all the while loops you've written whose last line is the increment statement?  It makes it a lot easier to see the loop at a high level without reading for detail.

(Also, this loop should probably be over Instruction*, rather than raw addresses.)